In [207]:
import pandas as pd 

In [208]:
data = pd.read_csv('C:Downloads/Data_Dictionary_for_CSV_2022_10_04.csv')

In [209]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30718 entries, 0 to 30717
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   data_source_title  30718 non-null  object
 1   report_type_title  30718 non-null  object
 2   field_title        30718 non-null  object
 3   sql_field_name     30718 non-null  object
 4   api_field_name     30718 non-null  object
dtypes: object(5)
memory usage: 1.2+ MB


In [210]:
data.columns 

Index(['data_source_title', 'report_type_title', 'field_title',
       'sql_field_name', 'api_field_name'],
      dtype='object')

In [212]:
data

data_source_title report_type_title            field_title  \
0                  ACME      Transactions    Transaction Item Id   
1                  ACME      Transactions         Transaction Id   
2                  ACME      Transactions               Order Id   
3                  ACME      Transactions          Membership Id   
4                  ACME      Transactions               Event Id   
...                 ...               ...                    ...   
30713       Zignal Labs  Newsroom Stories                  Title   
30714       Zignal Labs  Newsroom Stories                 Source   
30715       Zignal Labs  Newsroom Stories                    Url   
30716       Zignal Labs  Newsroom Stories              Sentiment   
30717       Zignal Labs  Newsroom Stories  Potential Impressions   

              sql_field_name         api_field_name  
0        transaction_item_id      TransactionItemId  
1             transaction_id          TransactionId  
2                   order_id                OrderId  
3              membership_id           MembershipId  
4                   event_id                EventId  
...                      ...                    ...  
30713                  title                  title  
30714                 source                 source  
30715                    url                    url  
30716              sentiment              sentiment  
30717  potential_impressions  potential_impressions  

[30718 rows x 5 columns]

In [213]:
data.data_source_title.value_counts()

Facebook                      5171
Google Universal Analytics    2789
Google Ads                    1676
Google Campaign Manager       1410
Adobe Analytics Data Feed     1084
                              ... 
Eventbrite                       7
Performance Marketing Lab        5
Example                          5
Business.Idealo                  4
Spotify Organic                  4
Name: data_source_title, Length: 177, dtype: int64

In [214]:
data.report_type_title.value_counts()

All Fields                1084
Ads                       1013
Campaign                   621
Geo                        403
Creatives                  395
                          ... 
Spam Complaint List          2
Tweet Entity                 2
Segment Values Mapping       2
Ads Target (Test2)           1
Custom contact actions       1
Name: report_type_title, Length: 943, dtype: int64

In [215]:
data.describe()

data_source_title report_type_title field_title sql_field_name  \
count              30718             30718       30718          30718   
unique               177               943        9180           8641   
top             Facebook        All Fields      Clicks    campaign_id   
freq                5171              1084         420            594   

       api_field_name  
count           30718  
unique           9296  
top               ???  
freq              810

In [216]:
data[data['data_source_title']=='ACME']

data_source_title report_type_title                    field_title  \
0               ACME      Transactions            Transaction Item Id   
1               ACME      Transactions                 Transaction Id   
2               ACME      Transactions                       Order Id   
3               ACME      Transactions                  Membership Id   
4               ACME      Transactions                       Event Id   
5               ACME      Transactions                  Discount Code   
6               ACME      Transactions                    Customer Id   
7               ACME      Transactions             Transaction Amount   
8               ACME      Transactions                       Quantity   
9               ACME      Transactions                 Payment Amount   
10              ACME      Transactions                  Order Balance   
11              ACME      Transactions                   Order Amount   
12              ACME      Transactions         Net Transaction Amount   
13              ACME      Transactions                     Net Amount   
14              ACME      Transactions          Discounted Unit Price   
15              ACME      Transactions  Discounted Transaction Amount   
16              ACME      Transactions              Discounted Amount   
17              ACME      Transactions                 Discount Value   
18              ACME      Transactions     Discount Transaction Value   
19              ACME      Transactions                         Amount   
20              ACME      Transactions                       Zip Code   
21              ACME      Transactions               Transaction Type   
22              ACME      Transactions               Transaction Date   
23              ACME      Transactions                    Ticket Type   
24              ACME      Transactions                      Sale Type   
25              ACME      Transactions                   Sale Channel   
26              ACME      Transactions                  Purchase Type   
27              ACME      Transactions                 Payment Status   
28              ACME      Transactions                 Payment Method   
29              ACME      Transactions               Payment Due Date   
30              ACME      Transactions          Parent Transaction Id   
31              ACME      Transactions                 OrderItem Type   
32              ACME      Transactions                   Order Number   
33              ACME      Transactions            Order Creation Date   
34              ACME      Transactions              Membership Number   
35              ACME      Transactions               Event Start Time   
36              ACME      Transactions                     Event Name   
37              ACME      Transactions                  Discount Name   
38              ACME      Transactions                           City   
39              ACME      Transactions                         Add On   

                   sql_field_name               api_field_name  
0             transaction_item_id            TransactionItemId  
1                  transaction_id                TransactionId  
2                        order_id                      OrderId  
3                   membership_id                 MembershipId  
4                        event_id                      EventId  
5                   discount_code                 DiscountCode  
6                     customer_id                   CustomerId  
7              transaction_amount            TransactionAmount  
8                        quantity                     Quantity  
9                  payment_amount                PaymentAmount  
10                  order_balance                 OrderBalance  
11                   order_amount                  OrderAmount  
12         net_transaction_amount         NetTransactionAmount  
13                     net_amount                    NetAmount  
14          discounted_unit_price 

In [217]:
data[data['data_source_title']=='Zignal Labs']

data_source_title report_type_title             field_title  \
30696       Zignal Labs       Key Metrics                     Mqs   
30697       Zignal Labs       Key Metrics          Total Mentions   
30698       Zignal Labs       Key Metrics              News Blogs   
30699       Zignal Labs       Key Metrics                  Social   
30700       Zignal Labs       Key Metrics  Impressions News Blogs   
30701       Zignal Labs       Key Metrics                Positive   
30702       Zignal Labs       Key Metrics                Negative   
30703       Zignal Labs       Key Metrics                 Neutral   
30704       Zignal Labs       Key Metrics                 Unknown   
30705       Zignal Labs  Newsroom Stories            Persisted At   
30706       Zignal Labs  Newsroom Stories                    Body   
30707       Zignal Labs  Newsroom Stories               Influence   
30708       Zignal Labs  Newsroom Stories                Story Id   
30709       Zignal Labs  Newsroom Stories                Language   
30710       Zignal Labs  Newsroom Stories                  Gender   
30711       Zignal Labs  Newsroom Stories                 Country   
30712       Zignal Labs  Newsroom Stories            Media Outlet   
30713       Zignal Labs  Newsroom Stories                   Title   
30714       Zignal Labs  Newsroom Stories                  Source   
30715       Zignal Labs  Newsroom Stories                     Url   
30716       Zignal Labs  Newsroom Stories               Sentiment   
30717       Zignal Labs  Newsroom Stories   Potential Impressions   

               sql_field_name                api_field_name  
30696                     mqs       mqs.media_quality_score  
30697          total_mentions                mentions.count  
30698              news_blogs              :calc.news_blogs  
30699                  social                  :calc.social  
30700  impressions_news_blogs  :calc.impressions_news_blogs  
30701                positive  sentiment.sentiment.positive  
30702                negative  sentiment.sentiment.negative  
30703                 neutral   sentiment.sentiment.neutral  
30704                 unknown   sentiment.sentiment.unknown  
30705            persisted_at                  persisted_at  
30706                    body                          body  
30707               influence                     influence  
30708                story_id                           _id  
30709                language                      language  
30710                  gender                        gender  
30711                 country                       country  
30712            media_outlet                  media_outlet  
30713                   title                         title  
30714                  source                        source  
30715                     url                           url  
30716               sentiment                     sentiment  
30717   potential_impressions         potential_impressions

In [218]:
id_list =[]
for i in range(len(data['report_type_title'])):
    if 'Metric' in data['report_type_title'][i] or 'Metrics' in data['report_type_title'][i]:
        id_list.append(i)

In [219]:
len(id_list)

107

In [220]:
data_1 = data.copy()

In [221]:
def delited(row):
    row = str(row).lower()
    row=str(row)
    row=row.replace('_', '')
    row=row.replace(' ', '')
    return row

In [222]:
#data_1['report_type_title'] = data_1['report_type_title'].apply(delited)
data_1['field_title'] = data_1['field_title'].apply(delited)
data_1['sql_field_name'] = data_1['sql_field_name'].apply(delited)
data_1['api_field_name'] = data_1['api_field_name'].apply(delited)

In [223]:
data_1

data_source_title report_type_title           field_title  \
0                  ACME      Transactions     transactionitemid   
1                  ACME      Transactions         transactionid   
2                  ACME      Transactions               orderid   
3                  ACME      Transactions          membershipid   
4                  ACME      Transactions               eventid   
...                 ...               ...                   ...   
30713       Zignal Labs  Newsroom Stories                 title   
30714       Zignal Labs  Newsroom Stories                source   
30715       Zignal Labs  Newsroom Stories                   url   
30716       Zignal Labs  Newsroom Stories             sentiment   
30717       Zignal Labs  Newsroom Stories  potentialimpressions   

             sql_field_name        api_field_name  
0         transactionitemid     transactionitemid  
1             transactionid         transactionid  
2                   orderid               orderid  
3              membershipid          membershipid  
4                   eventid               eventid  
...                     ...                   ...  
30713                 title                 title  
30714                source                source  
30715                   url                   url  
30716             sentiment             sentiment  
30717  potentialimpressions  potentialimpressions  

[30718 rows x 5 columns]

In [224]:
data_1['field'] = data_1['field_title']==data_1['sql_field_name']

In [225]:
data_1['field'].value_counts()

True     20960
False     9758
Name: field, dtype: int64

In [226]:
data_1[data_1['field']==False]

data_source_title         report_type_title  \
61       ActiveCampaign         Campaign Messages   
95               Act on            Message Report   
117              Act on         Message Drilldown   
121              Act on         Message Drilldown   
135              Adform        Device Performance   
...                 ...                       ...   
30577           Zemanta  Content Ads By Countries   
30578           Zemanta  Content Ads By Countries   
30579           Zemanta  Content Ads By Countries   
30582           Zemanta  Content Ads By Countries   
30583           Zemanta  Content Ads By Countries   

                           field_title                      sql_field_name  \
61                            campaign                        campaignname   
95                       uniqueclicked                       unuqieclicked   
117                               when                               swhen   
121                          ipaddress                            ipadress   
135            trackedads(impressions)                         impressions   
...                                ...                                 ...   
30577     %nativemeasurableimpressions  nativemeasurableimpressionspercent   
30578    nativenot-viewableimpressions        nativenotviewableimpressions   
30579  nativenot-measurableimpressions      nativenotmeasurableimpressions   
30582                          avg.cpm                              avgcpm   
30583                          avg.cpc                              avgcpc   

                                     api_field_name  field  
61                                   campaigns.name  False  
95                                        clicked.0  False  
117                                            when  False  
121                                              ip  False  
135    impressionsdatasource.adformaduniqueness.all  False  
...                                             ...    ...  
30577                  %nativemeasurableimpressions  False  
30578                 nativenot-viewableimpressions  False  
30579               nativenot-measurableimpressions  False  
30582                                       avg.cpm  False  
30583                                       avg.cpc  False  

[9758 rows x 6 columns]

In [227]:
data_1['report_type_title'].value_counts()

All Fields                1084
Ads                       1013
Campaign                   621
Geo                        403
Creatives                  395
                          ... 
Spam Complaint List          2
Tweet Entity                 2
Segment Values Mapping       2
Ads Target (Test2)           1
Custom contact actions       1
Name: report_type_title, Length: 943, dtype: int64

In [228]:
data_1.describe()

data_source_title report_type_title field_title sql_field_name  \
count              30718             30718       30718          30718   
unique               177               943        8570           8579   
top             Facebook        All Fields  campaignid     campaignid   
freq                5171              1084         597            594   

       api_field_name  field  
count           30718  30718  
unique           8723      2  
top               ???   True  
freq              810  20960

In [229]:
data 

data_source_title report_type_title            field_title  \
0                  ACME      Transactions    Transaction Item Id   
1                  ACME      Transactions         Transaction Id   
2                  ACME      Transactions               Order Id   
3                  ACME      Transactions          Membership Id   
4                  ACME      Transactions               Event Id   
...                 ...               ...                    ...   
30713       Zignal Labs  Newsroom Stories                  Title   
30714       Zignal Labs  Newsroom Stories                 Source   
30715       Zignal Labs  Newsroom Stories                    Url   
30716       Zignal Labs  Newsroom Stories              Sentiment   
30717       Zignal Labs  Newsroom Stories  Potential Impressions   

              sql_field_name         api_field_name  
0        transaction_item_id      TransactionItemId  
1             transaction_id          TransactionId  
2                   order_id                OrderId  
3              membership_id           MembershipId  
4                   event_id                EventId  
...                      ...                    ...  
30713                  title                  title  
30714                 source                 source  
30715                    url                    url  
30716              sentiment              sentiment  
30717  potential_impressions  potential_impressions  

[30718 rows x 5 columns]

In [230]:
data[data['data_source_title']=='ACME']

data_source_title report_type_title                    field_title  \
0               ACME      Transactions            Transaction Item Id   
1               ACME      Transactions                 Transaction Id   
2               ACME      Transactions                       Order Id   
3               ACME      Transactions                  Membership Id   
4               ACME      Transactions                       Event Id   
5               ACME      Transactions                  Discount Code   
6               ACME      Transactions                    Customer Id   
7               ACME      Transactions             Transaction Amount   
8               ACME      Transactions                       Quantity   
9               ACME      Transactions                 Payment Amount   
10              ACME      Transactions                  Order Balance   
11              ACME      Transactions                   Order Amount   
12              ACME      Transactions         Net Transaction Amount   
13              ACME      Transactions                     Net Amount   
14              ACME      Transactions          Discounted Unit Price   
15              ACME      Transactions  Discounted Transaction Amount   
16              ACME      Transactions              Discounted Amount   
17              ACME      Transactions                 Discount Value   
18              ACME      Transactions     Discount Transaction Value   
19              ACME      Transactions                         Amount   
20              ACME      Transactions                       Zip Code   
21              ACME      Transactions               Transaction Type   
22              ACME      Transactions               Transaction Date   
23              ACME      Transactions                    Ticket Type   
24              ACME      Transactions                      Sale Type   
25              ACME      Transactions                   Sale Channel   
26              ACME      Transactions                  Purchase Type   
27              ACME      Transactions                 Payment Status   
28              ACME      Transactions                 Payment Method   
29              ACME      Transactions               Payment Due Date   
30              ACME      Transactions          Parent Transaction Id   
31              ACME      Transactions                 OrderItem Type   
32              ACME      Transactions                   Order Number   
33              ACME      Transactions            Order Creation Date   
34              ACME      Transactions              Membership Number   
35              ACME      Transactions               Event Start Time   
36              ACME      Transactions                     Event Name   
37              ACME      Transactions                  Discount Name   
38              ACME      Transactions                           City   
39              ACME      Transactions                         Add On   

                   sql_field_name               api_field_name  
0             transaction_item_id            TransactionItemId  
1                  transaction_id                TransactionId  
2                        order_id                      OrderId  
3                   membership_id                 MembershipId  
4                        event_id                      EventId  
5                   discount_code                 DiscountCode  
6                     customer_id                   CustomerId  
7              transaction_amount            TransactionAmount  
8                        quantity                     Quantity  
9                  payment_amount                PaymentAmount  
10                  order_balance                 OrderBalance  
11                   order_amount                  OrderAmount  
12         net_transaction_amount         NetTransactionAmount  
13                     net_amount                    NetAmount  
14          discounted_unit_price 

In [231]:
data.pivot_table(index=['data_source_title'], values='field_title', aggfunc='count')

field_title
data_source_title             
ACME                        40
AWIN                        70
Act on                      48
ActiveCampaign              46
AdFox                       43
...                        ...
Youtube Organic            165
Zemanta                    247
Zendesk                    104
Zignal Labs                 22
eBay                        31

[177 rows x 1 columns]

In [232]:
data.groupby(['report_type_title']).agg('count')

data_source_title  field_title  \
report_type_title                                                
Account Adset Report Type                      19           19   
Account Geo                                    17           17   
Account Month Reach                            34           34   
Account Monthly                                72           72   
Account Platform Report Type                   19           19   
...                                           ...          ...   
statistics                                     12           12   
unstable                                        2            2   
Сampaign Event                                  7            7   
Сonversions                                    21           21   
Сreative Сustom Events Country                 31           31   

                                sql_field_name  api_field_name  
report_type_title                                               
Account Adset Report Type                   19              19  
Account Geo                                 17              17  
Account Month Reach                         34              34  
Account Monthly                             72              72  
Account Platform Report Type                19              19  
...                                        ...             ...  
statistics                                  12              12  
unstable                                     2               2  
Сampaign Event                               7               7  
Сonversions                                 21              21  
Сreative Сustom Events Country              31              31  

[943 rows x 4 columns]

In [233]:
data 

data_source_title report_type_title            field_title  \
0                  ACME      Transactions    Transaction Item Id   
1                  ACME      Transactions         Transaction Id   
2                  ACME      Transactions               Order Id   
3                  ACME      Transactions          Membership Id   
4                  ACME      Transactions               Event Id   
...                 ...               ...                    ...   
30713       Zignal Labs  Newsroom Stories                  Title   
30714       Zignal Labs  Newsroom Stories                 Source   
30715       Zignal Labs  Newsroom Stories                    Url   
30716       Zignal Labs  Newsroom Stories              Sentiment   
30717       Zignal Labs  Newsroom Stories  Potential Impressions   

              sql_field_name         api_field_name  
0        transaction_item_id      TransactionItemId  
1             transaction_id          TransactionId  
2                   order_id                OrderId  
3              membership_id           MembershipId  
4                   event_id                EventId  
...                      ...                    ...  
30713                  title                  title  
30714                 source                 source  
30715                    url                    url  
30716              sentiment              sentiment  
30717  potential_impressions  potential_impressions  

[30718 rows x 5 columns]

After studying the data, the idea came up to make a list with all the metrics.

Source : https://habr.com/ru/post/675722/

In [235]:
data['field_title'] = data['field_title'].str.lower()
data['report_type_title'] = data['report_type_title'].str.lower()
set_1 = set(data['field_title'])

In [236]:
metrics = [
    'Page loading speed',
    'Total user',
    'Total visit',
    'Returning User Rate',
    'Time Spent on Site',
    'Pages per Visit', 
    'Pages per Visit',
    'Bounce Rate', 
    'On-Site Engagement Activity',
    'On-Site Engagement Activity',
    'Shopping Cart Abandonment Rate',
    'Leads',
    'Marketing Qualified Leads',
    'Sales Qualified Lead' ,
    'Product Qualified Lead',
    'Conversion Rate',
    'Lead Close Rate',
    'Sales',
    'Conversions',
    'Transactions',
    'Associated Conversions',
    'Engagement Rate',
    'Engagement Rate by Reach',
    'Organic Followers',
    'Audience Growth Rate',
    'Amplification Rate',
    'Love Rate',
    'Talk Rate',
    'Negative Feedback',
    'Response Time',
    'Delivery Rate' ,
    'Open Rate',
    'Click To Open Rate' ,
    'Unsubscribe Rate' ,
    'Complaint Rate' ,
    'View',
    'Click',
    'Reach',
    'Frequency Display',
    'Click-Through Rate',
    'Cost per Click',
    'Cost per Visitor',
    'Cost per View ',
    'Cost per Mille',
    'Cost per Action',
    'Cost per Lead',
    'Leads Generated per Offer',
    'Cost per Install',
    'Cost per Fan',
    'Cost per Sale',
    'Return on Ad Spend',
    'Purchase Frequency' ,
    'Repeat Purchase Rate',
    'Order Gap Analysis',
    'Revenue Churn',
    'Total Revenue',
    'Cost of Good Sale',
    'Average Order Value',
    'Margin',
    'Customer Acquisition Cost',
    'Average Revenue per User',
    'Average Revenue per Paying User',
    'Average Revenue per Paying User',
    'Monthly Recurring Revenue',
    'Lifetime Value',
    'Return on Marketing Investment',
    'Return on Marketing Investment',
    'Return on Investment'
] #Source https://habr.com/ru/post/675722/
for i in range(len(metrics)):
    metrics[i] = metrics[i].lower()

In [237]:
metrics_names = []
for i in set_1:
    if str(i) in metrics:
        metrics_names.append(i) 
metrics_names

['bounce rate',
 'leads',
 'click',
 'negative feedback',
 'conversion rate',
 'total revenue',
 'transactions',
 'average order value',
 'reach',
 'engagement rate',
 'conversions',
 'sales',
 'open rate']

In [238]:
metrics_names_1 = ['user',
                  'users',
                  'visit',
                  'visits',
                  'rate', 
                  'bounce rate',
                 'leads',
                 'click',
                 'negative feedback',
                 'conversion rate',
                 'total revenue',
                 'transactions',
                 'average order value',
                 'reach',
                 'engagement rate',
                 'conversions',
                 'sales',
                 'open rate',
                  'total',
                  'sales',
                  'View',
                'clicks',
                'reach',
                  'cost',
                  'margin',
                  'average', 
                  'return',
                  'sale',
                  'subscribed']

In [239]:
cnd=0
for i in data['field_title']:
    for j in metrics_names_1:
        if j in i: 
            cnd+=1
        #print(i) click

In [240]:
cnd

10210

In [241]:
cnd/len(data) *100

33.23784100527378

After that, I checked the "field_title" column for the presence of a metric from the list "metrics_names_1".

If there is a metric, we increase the counter by 1. 

It turned out 10210 lines. This is ~33%